compiler/rustc_const_eval/src/interpret/cast.rs

```diff
@@ -336,6 +389,24 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                     );
                 }
 
+                let just_metadata = self.tcx.lang_items().just_metadata();
+                if Some(def_a.did()) == just_metadata && Some(def_b.did()) == just_metadata {
+                    let cast_ty_field = cast_ty.field(self, 0);
+                    if cast_ty_field.is_zst() {
+                        return Ok(());
+                    }
+                    let a_pointee = substs_a.type_at(0);
+                    let b_pointee = substs_b.type_at(0);
+                    let src_field = self.operand_field(src, 0)?;
+                    let dst_field = self.place_field(dest, 0)?;
+                    if src_field.layout.ty == cast_ty_field.ty {
```

Comment:
Is this branch needed here?
It is needed for the generic struct case since there we go over all fields of the struct one-by-one, and then only the last one is actually being unsized. (Probably the better way to do that would be to use the index `i` rather than `ty` equality to distinguish these cases.)

Reply:
It actually isn't necessarily the case I *think*, or perhaps the ICE error spans are just misleading: when originally implementing it seemed that this code path gets called for identity as well (e.g. `JustMetadata<u8> -> JustMetadata<u8>`. I'll double check that tomorrow.

Reply:
>  it seemed that this code path gets called for identity as well (e.g. JustMetadata<u8> -> JustMetadata<u8>

Oh, strange. In that case please add a comment explaining that.

Reply:
I was incorrect: this codepath isn't taken in that case, and replacing the positive branch here with a `bug!()` doesn't ICE. As such, I've removed it 😄

The "cast not implemented" ICE span is not super helpful, thus what got me thinking that this codepath was taken even for the identity case.